lib/Backends/BackendUtils.cpp

```diff
@@ -81,6 +74,55 @@ runtime::RuntimeBundle::getSymbolInfo(const Named *v) const {
   return it->second;
 }
 
+/// If \p PH is an output placeholder, \returns true.
+/// This is determined by checking if the PH has a user which uses the PH as an
+/// overwritten input.
+bool isOutput(const Placeholder *PH) {
+  // Check if trainable, if it is return true.
+  if (PH->isTraining()) {
+    return true;
+  }
+  for (auto &use : PH->getUsers()) {
+    // Look through the inputs of the PH's users. If an input is overwritten
+    // check if it's the PH, if it is return true.
+    for (unsigned i = 0; i < use.getUser()->getNumInputs(); i++) {
+      // If the input is not overwritten we can continue.
+      if (!use.getUser()->isOverwrittenNthInput(i)) {
+        continue;
+      }
+      auto input = use.getUser()->getNthInput(i);
+      if (input == PH) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+/// If \p PH is an input placeholder, \returns true.
+bool isInput(const Placeholder *PH) {
+  // Check if trainable, if it is return true.
+  if (PH->isTraining()) {
+    return true;
+  }
+  for (auto &use : PH->getUsers()) {
+    // Check if PH is a non overwritten input to a saveNode.
+    if (auto save = llvm::dyn_cast<SaveNode>(use.getUser())) {
+      auto input = save->getInput();
+      // Check if the PH is the input to the saveNode.
+      if (input != PH) {
+        continue;
+      }
+      // Check if input is overwritten.
```

Comment:
What is operand zero?  Prefer to use the descriptive label of SaveNode. (getDestination I think)

Reply:
Operand 0 is input and 1 is output for a saveNode, but after looking at isOverWrittenNthInput, this check is useless, since isOverWritten returns false for all values except 1.